File: Spark/simple-spark/src/main/java/org/gb/sample/spark/income/IncomeBandMapper.java
package org.gb.sample.spark.income;

import java.io.Serializable;
import java.util.List;
import java.util.Map;

import org.apache.spark.api.java.JavaRDD;

public class IncomeBandMapper implements Serializable {

	private static final long serialVersionUID = 1272508542730740489L;

	Map<Band, Long> getCountPerIncomeBand(JavaRDD<String> nameVsIncomeLines, List<Integer> incomeSlabs) {
		JavaRDD<Band> incomeBands = nameVsIncomeLines.map(nameVsIncomeLine -> nameVsIncomeLine.split(","))
				.filter(tokenizedNameVsIncomeLine -> tokenizedNameVsIncomeLine.length == 3)
				.map(tokenizedNameVsIncomeLine -> mapToBand(tokenizedNameVsIncomeLine[2], incomeSlabs))
				.filter(band -> band != null);
		int noOfPartitions = incomeBands.rdd().getNumPartitions();
		System.out.printf("No. of partitions: %d.\n", noOfPartitions);
		Map<Band, Long> countPerIncomeBand = nameVsIncomeLines.map(nameVsIncomeLine -> nameVsIncomeLine.split(","))
				.filter(tokenizedNameVsIncomeLine -> tokenizedNameVsIncomeLine.length == 3)
				.map(tokenizedNameVsIncomeLine -> mapToBand(tokenizedNameVsIncomeLine[2], incomeSlabs))
				.filter(band -> band != null).countByValue();
		return countPerIncomeBand;
	}

	private Band mapToBand(String incomeStr, List<Integer> incomeSlabs) {
		Integer income = null;
		try {
			income = Integer.parseInt(incomeStr);
		} catch (NumberFormatException nfe) {
			System.err.println(String.format("incomeStr: %s could not be converted to an integer", incomeStr));
			return null;
		}

		Band band = null;
		for (int i = 0; i < incomeSlabs.size(); i++) {
			Integer currentSlab = incomeSlabs.get(i);
			Integer nextSlab = null;

			if ((i + 1) < incomeSlabs.size()) {
				nextSlab = incomeSlabs.get(i + 1);
			}

			if (nextSlab == null) {
				if (income.intValue() > currentSlab.intValue()) {
					band = new Band(currentSlab, nextSlab);
					break;
				}
			} else if (income.intValue() > currentSlab.intValue() && income.intValue() <= nextSlab.intValue()) {
				band = new Band(currentSlab, nextSlab);
				break;
			}
		}
		System.out.printf("Income: %d falls in band: %s.\n", income.intValue(), band);
		return band;
	}

}

File: Spark/simple-spark-scala/src/main/scala/org/gb/sample/spark/income/IncomeBandMapper.scala
package org.gb.sample.spark.income

import org.apache.spark.rdd.RDD
import scala.util.control.Breaks._

class IncomeBandMapper(nameVsIncomeLines: RDD[String], incomeSlabs: List[Int]) extends Serializable {
    def getCountPerIncomeBand(): Map[Band, Long] = {
        val countPerIncomeBand = nameVsIncomeLines.map(_.split(",")).filter(_.length == 3).map(_(2)).map(mapToBand).filter(_ != null).countByValue()
        countPerIncomeBand.toMap
    }

    private def mapToBand(incomeStr: String): Band = {
        var income = 0
        try {
            income = incomeStr.toInt
        } catch {
            case nfe: NumberFormatException => return null
        }

        var band: Band = null;
        breakable {
            for (i <- 0 to incomeSlabs.size) {
                val currentSlab = incomeSlabs(i)
                val nextSlab = if ((i + 1) < incomeSlabs.size) incomeSlabs(i + 1) else -1

                if (nextSlab == -1) {
                    if (income > currentSlab) {
                        band = Band(currentSlab, nextSlab)
                        break;
                    }
                } else if (income > currentSlab && income <= nextSlab) {
                    band = Band(currentSlab, nextSlab)
                    break;
                }
            }
        }
        println(s"Income: $income falls in band: $band.")
        return band
    }
}